File: ThisIsCodingTest/이것이코딩테스트다_15장_가사검색.py
import unittest


class Node:
    def __init__(self, char, data=None):
        self.char = char
        self.length = []
        self.data = data
        self.children = dict()


class Trie:
    def __init__(self):
        self.head = Node(None)

    def insert(self, word):
        cur = self.head

        for idx, char in enumerate(word):
            if char not in cur.children:
                cur.children[char] = Node(char)
            cur.children[char].length.append(len(word))
            cur = cur.children[char]

        cur.data = word

    def starts_with_prefix(self, prefix):
        cur = self.head

        for idx, char in enumerate(prefix):
            if char != "?":
                if char not in cur.children:
                    return 0

                cur = cur.children[char]
                continue

            return cur.length.count(len(prefix))


def solution(words, queries):
    answer = []
    words_length = []

    trie = Trie()
    reverse_trie = Trie()
    for idx, word in enumerate(words):
        trie.insert(word)
        reverse_trie.insert(word[::-1])
        words_length.append(len(word))

    for idx, query in enumerate(queries):
        if query[0] != "?":
            answer.append(trie.starts_with_prefix(query))
            continue

        if query[-1] != "?":
            answer.append(reverse_trie.starts_with_prefix(query[::-1]))
            continue

        answer.append(words_length.count(len(query)))

    return answer


class TestStringMethods(unittest.TestCase):
    def test_example_input(self):
        result = solution(["frodo", "front", "frost", "frozen", "frame", "kakao"], ["fro??", "????o", "fr???", "fro???", "pro?"])
        self.assertEqual(result, [3, 2, 4, 1, 0])

    def test_reversed_word(self):
        result = solution(["ab"], ["?a"])
        self.assertEqual(result, [0])


if __name__ == '__main__':
    unittest.main()

# 참고
# https://velog.io/@hope1213/프로그래머스-가사검색-파이썬

File: this-is-coding-test/이것이코딩테스트다_15장_가사검색.js
class Node {
  constructor(data = null) {
    this.data = data;
    this.len = 0;
    this.children = {};
  }
}

class Trie {
  constructor() {
    this.head = new Node();
  }

  insert(word) {
    let node = this.head;
    node.len += 1;

    for (const w of word) {
      if (!node.children.hasOwnProperty(w)) {
        node.children[w] = new Node();
      }

      node = node.children[w];
      node.len += 1;
    }

    node.data = word;
  }

  startWithPrefix(prefix) {
    let node = this.head;
    
    for (const p of prefix) {
      if (p === "?") {
        break;
      }

      if (!node.children.hasOwnProperty(p)) {
        return 0;
      }

      node = node.children[p];
    }

    return node.len;
  }
}

const reverseWord = (word) => {
  let reversedWord = '';

  for (const w of word) {
    reversedWord = w + reversedWord;
  }

  return reverseWord;
}

const solution = (words, queries) => {
  const tries = {};
  const reversedTries = {};

  // const trie = new Trie();
  // const reversedTrie = new Trie();

  words.forEach(word => {
    const wordLength = word.length;

    if (!tries.hasOwnProperty(wordLength.toString())) {
      tries[wordLength] = new Trie();
    }

    tries[wordLength].insert(word)

    const reversedWord = word.split('').reverse().join('');

    if (!reversedTries.hasOwnProperty(wordLength.toString())) {
      reversedTries[wordLength] = new Trie();
    }

    reversedTries[wordLength].insert(reversedWord);
  });

  return queries.map(query => {
    const queryLength = query.length;

    if (query[0] === "?") {
      if (!reversedTries.hasOwnProperty(queryLength)) {
        return 0;
      }

      return reversedTries[queryLength].startWithPrefix(query.split('').reverse().join(''));
    }
    
    if (!tries.hasOwnProperty(queryLength)) {
      return 0;
    }

    return tries[queryLength].startWithPrefix(query);
  });
}
